src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1184,20 +1185,44 @@ private void LookupMembersInTypeParameter(LookupResult current, TypeParameterSym
             LookupMembersInInterfacesWithoutInheritance(current, typeParameter.AllEffectiveInterfacesWithDefinitionUseSiteDiagnostics(ref useSiteInfo), name, arity, basesBeingResolved: null, options, originalBinder, typeParameter, diagnose, ref useSiteInfo);
         }
 
-        private static bool IsDerivedType(NamedTypeSymbol baseType, NamedTypeSymbol derivedType, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        private static bool IsDerivedType(NamedTypeSymbol baseType, NamedTypeSymbol derivedType, ConsList<TypeSymbol> basesBeingResolved, CSharpCompilation compilation, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
         {
             Debug.Assert(!TypeSymbol.Equals(baseType, derivedType, TypeCompareKind.ConsiderEverything2));
-            for (NamedTypeSymbol b = derivedType.BaseTypeWithDefinitionUseSiteDiagnostics(ref useSiteInfo); (object)b != null; b = b.BaseTypeWithDefinitionUseSiteDiagnostics(ref useSiteInfo))
+
+            if (basesBeingResolved?.Any() != true)
+            {
+                for (NamedTypeSymbol b = derivedType.BaseTypeWithDefinitionUseSiteDiagnostics(ref useSiteInfo); (object)b != null; b = b.BaseTypeWithDefinitionUseSiteDiagnostics(ref useSiteInfo))
+                {
+                    if (TypeSymbol.Equals(b, baseType, TypeCompareKind.ConsiderEverything2)) return true;
+                }
+            }
+            else
             {
-                if (TypeSymbol.Equals(b, baseType, TypeCompareKind.ConsiderEverything2)) return true;
+                PooledHashSet<NamedTypeSymbol> visited = null;
+
+                for (var b = (NamedTypeSymbol)derivedType.GetNextBaseTypeNoUseSiteDiagnostics(basesBeingResolved, compilation, ref visited);
+                     (object)b != null;
+                     b = (NamedTypeSymbol)b.GetNextBaseTypeNoUseSiteDiagnostics(basesBeingResolved, compilation, ref visited))
+                {
+                    b.OriginalDefinition.AddUseSiteInfo(ref useSiteInfo);
+
+                    if (TypeSymbol.Equals(b, baseType, TypeCompareKind.ConsiderEverything2))
```

Comment:
Can we avoid introducing new usage of `ConsiderEverything2`?

Reply:
> Can we avoid introducing new usage of `ConsiderEverything2`?

This line of code is a clone of an existing line in the same function. Figuring out appropriate mode of comparison and adding related tests was outside of scope of this PR.